src/dapp/libexec/dapp/dapp-build

```diff
@@ -38,58 +48,81 @@ info() {
-
 # use a custom path is DAPP_SOLC is set
 SOLC=${DAPP_SOLC:-solc}
 
-$SOLC --help | grep -q -- --storage-layout && json_opts+=,storage-layout
-$SOLC --help | grep -q -- --overwrite && opts+=(--overwrite)
+if [ "$DAPP_LINK_TEST_LIBRARIES" = 1 ]; then
+    sender=0x00a329c0648769a73afac7f9381e08fb43dbea72
+    nonce=48576
+    mapfile -t libraries < <(dapp --find-libraries)
 
-remappings=$(dapp remappings)
-if [[ $remappings ]]; then
-  while read -r line; do opts+=("$line"); done <<<"$remappings"
-fi
+    if [[ ${#libraries[@]} -eq 0 ]]; then exit; fi
 
-if [ "$DAPP_LINK_TEST_LIBRARIES" = 1 ]; then
-  DAPP_LINK_TEST_LIBRARIES=0 dapp build
-  sender=0x00a329c0648769a73afac7f9381e08fb43dbea72
-  nonce=48576
-  mapfile -t libraries < <(dapp --find-libraries)
-
-  if [[ ${#libraries[@]} -eq 0 ]]; then exit; fi
-
-  info "rebuilding with linked libraries"
-  for lib in "${libraries[@]}"; do
-    address=$(dapp address "$sender" "$nonce")
-    links+=" $lib:$address"
-    ((nonce++))
-  done
-  opts+=(--libraries "$links")
+    info "building with linked libraries"
+    for lib in "${libraries[@]}"; do
+        address=$(dapp address "$sender" "$nonce")
+        links+=" $lib:$address"
+        ((nonce++))
+    done
 else
-  (set -x; dapp clean)
+    (set -x; dapp clean)
 fi
 
+remappings=$(dapp remappings)
+
+export DAPP_LIBRARIES=${DAPP_LIBRARIES-"$links"}
+export DAPP_REMAPPINGS=${DAPP_REMAPPINGS-"$remappings"}
+export DAPP_BUILD_OPTIMIZE
+
 mkdir -p "$DAPP_OUT"
-mapfile -t files < <(find "${DAPP_SRC}" -name '*.sol')
-(set -x; $SOLC "${opts[@]}" "${json_opts[@]}" /=/ "${files[@]}" > "${DAPP_JSON}")
-
-if [ "$DAPP_BUILD_EXTRACT" ]; then
-  mapfile -t contracts < <(<"$DAPP_JSON" jq '.contracts|keys[]' -r | sort -u -t: -k2 | sort)
-  data=$(<"$DAPP_JSON" jq '.contracts' -r)
-  count=1
-  total=${#contracts[@]}
-  for path in "${contracts[@]}"; do
-    info -ne "Extracting build data... [$count/$total]\\r"
-    ((count++))
-    name="${path#*:}"
-    contract=$(echo "$data" | jq '.[''"'"$path"'"'']')
-    echo "$contract" | jq '.["abi"]' -r > "$DAPP_OUT/$name.abi"
-    echo "$contract" | jq '.["bin"]' -r > "$DAPP_OUT/$name.bin"
-    echo "$contract" | jq '.["bin-runtime"]' -r > "$DAPP_OUT/$name.bin-runtime"
-    echo "$contract" | jq '.["metadata"]' -r > "$DAPP_OUT/$name.metadata"
-  done
-  echo
-  info "warning: --extract is slow and may have filename collisions. All build data can be found in $DAPP_JSON.\\n"
+
+if [[ ! "$DAPP_BUILD_LEGACY" = 1 ]]; then
+
+    if [[ "$DAPP_STANDARD_JSON" ]]; then
+        $<("$DAPP_STANDARD_JSON") | "$SOLC" --standard-json --allow-paths . > "$DAPP_JSON"
+    else
+        dapp mk-standard-json | "$SOLC" --standard-json --allow-paths . > "$DAPP_JSON"
+    fi
+
+    # pipe errors to stderr
+    jq -r '.errors | map(."formattedMessage") | @sh' "$DAPP_JSON" | sed -e "s/' //g" -e "s/'//g" 1>&2
+    # if sources is empty, the compilation failed
+    [[ $(jq -r '.sources' "$DAPP_JSON") = "{}" ]] && exit 1 || exit 0
+
+else
```

Comment:
Let's put this in a separate `dapp---build-legacy` file

Reply:
Oh and add a loud deprecation warning so we can delete it :)

Reply:
added